File: Interpreter_TS/interpreter.ts
const EOF = 'EOF'
const INTEGER = 'INTEGER'
const PLUS = 'PLUS'
const MINUS = 'MINUS'
const MUL = 'MUL'
const DIV = 'DIV'
const LPAREN = 'LPAREN'
const RPAREN = 'RPAREN'

declare interface String {
  isspace(): boolean;
  isdigit(): boolean;
}

/** 空格换行等 */
String.prototype.isspace = function () {
  return /\s/g.test((this as string))
}
/** 数字 */
String.prototype.isdigit = function () {
  return /^\d+$/.test((this as string))
}

class Token {
  type: string
  value: string | number | null
  constructor(type: string, value: string | number | null) {
    this.type = type
    this.value = value
  }
}
/**
 * 词法分析器
 */
class Lexer {
  text: string
  currentToken: string
  pos: number
  currentChar: string
  constructor(text: string) {
    this.text = text
    this.pos = 0
    this.currentToken = ''
    this.currentChar = this.text[this.pos]
  }
  number() {
    let result = ''
    while (this.currentChar != undefined && this.currentChar.isdigit()) {
      result += this.currentChar
      this.advance()
    }
    return new Token(INTEGER, Number(result))
  }
  advance() {
    this.pos += 1
    if (this.currentChar != undefined) {
      this.currentChar = this.text[this.pos]
    }
  }
  skipWhitespace() {
    while (this.currentChar != undefined && this.currentChar.isspace()) {
      this.advance()
    }
  }
  error() {
    let message = `lexer error currentChar '${this.currentChar}' pos ${this.pos}`
    throw new Error(message)
  }

  getNextToken() {
    while (this.currentChar != undefined) {
      if (this.currentChar.isspace()) {
        this.skipWhitespace()
        continue
      }
      if (this.currentChar.isdigit()) {
        return this.number()
      }
      if (this.currentChar == '/') {
        this.advance()
        return new Token(DIV, '/')
      }
      if (this.currentChar == '+') {
        this.advance()
        return new Token(PLUS, '+')
      }
      if (this.currentChar == '-') {
        this.advance()
        return new Token(MINUS, '-')
      }
      if (this.currentChar == '*') {
        this.advance()
        return new Token(MUL, '*')
      }
      if (this.currentChar == '(') {
        this.advance()
        return new Token(LPAREN, '(')
      }
      if (this.currentChar == ')') {
        this.advance()
        return new Token(RPAREN, ')')
      }
      this.error()
    }
    return new Token(EOF, null)
  }
}


class AST { }

class Num extends AST {
  token: Token
  value: string | number | null
  constructor(token: Token) {
    super()
    this.token = token
    this.value = token.value
  }
}

class BinOp extends AST {
  left: AST
  right: AST
  token: Token
  operator: Token
  constructor(left: AST, operator: Token, right: AST) {
    super()
    this.left = left
    this.token = this.operator = operator
    this.right = right
  }
}

/**
 * 语法分析器
 */
class Parser {
  lexer: Lexer
  currentToken: Token
  constructor(lexer: Lexer) {
    this.lexer = lexer
    this.currentToken = this.lexer.getNextToken()
  }
  eat(tokenType: string) {
    if (this.currentToken.type == tokenType) {
      this.currentToken = this.lexer.getNextToken()
    } else {
      this.error()
    }
  }
  error() {
    throw new Error('Error parsing input currentToken value: ' + this.currentToken.value)
  }
  factor() {
    /**
    factor : INTEGER | LPAREN expr RPAREN
    */
    let token = this.currentToken
    if (token.type == INTEGER) {
      this.eat(INTEGER)
      return new Num(token)
    } else if (token.type == LPAREN) {
      this.eat(LPAREN)
      let node = this.expr()
      this.eat(RPAREN)
      return node
    }
  }
  term() {
    // term : factor ((MUL | DIV) factor)*
    let node = this.factor()
    while ([MUL, DIV].includes(this.currentToken.type)) {
      let token = this.currentToken
      if (token.type == MUL) {
        this.eat(MUL)
      } else if (token.type == DIV) {
        this.eat(DIV)
      }
      node = new BinOp(node, token, this.factor())
    }
    return node
  }
  expr() {
    /**
     expr   : term ((PLUS | MINUS) term)*
     term   : factor ((MUL | DIV) factor)*
     factor : INTEGER | LPAREN expr RPAREN
     */
    let node = this.term()
    while ([PLUS, MINUS].includes(this.currentToken.type)) {
      let token = this.currentToken
      if (token.type == PLUS) {
        this.eat(PLUS)
      } else if (token.type == MINUS) {
        this.eat(MINUS)
      }
      node = new BinOp(node, token, this.term())
    }
    return node
  }
  parse() {
    return this.expr()
  }
}


class NodeVisitor {
  visit(node: AST) {
    let methodName = 'visit' + node.constructor.name
    if (this[methodName] === undefined) {
      throw new Error(`visit ${methodName} is not a functiontion`)
    }
    return this[methodName](node)
  }
}

/**
 * 解释器  访问者模式 
 */
class Interpreter extends NodeVisitor {
  tree: AST
  constructor(tree: AST) {
    super()
    this.tree = tree
  }
  visitBinOp(node: BinOp) {
    let visit = this.visit.bind(this)
    if (node.operator.type == PLUS) {
      return visit(node.left) + visit(node.right)
    }
    if (node.operator.type == MINUS) {
      return visit(node.left) - visit(node.right)
    }
    if (node.operator.type == MUL) {
      return visit(node.left) * visit(node.right)
    }
    if (node.operator.type == DIV) {
      return visit(node.left) / visit(node.right)
    }
  }
  visitNum(node: Num) {
    return node.value
  }
  interpret() {
    let tree = this.tree
    // console.log('tree', tree)
    if (tree == undefined) {
      return ''
    }
    return this.visit(tree)
  }
}


function ensure(condition: boolean, message: string) {
  if (!condition) {
    console.log("*** 测试失败:", message)
  } else {

    console.log("*** 测试成功:", message)
  }
}

function makeInterpreter(text: string) {
  let lexer = new Lexer(text)
  let parser = new Parser(lexer)
  let tree = parser.parse()
  let interpreter = new Interpreter(tree)
  return interpreter
}

function testExpression1() {
  let interpreter = makeInterpreter("3")
  let result = interpreter.interpret()
  ensure(result === 3, "test_expression1")
}

function testExpression2() {
  let expression = "2 + 7 * 4"
  let interpreter = makeInterpreter(expression)
  let result = interpreter.interpret()
  ensure(result === 30, expression)
}

function testExpression3() {
  let expression = "7 - 8 / 4"
  let interpreter = makeInterpreter(expression)
  let result = interpreter.interpret()
  ensure(result === 5, expression)
}

function testExpression4() {
  let expression = "14 + 2 * 3 - 6 / 2"
  let interpreter = makeInterpreter(expression)
  let result = interpreter.interpret()
  ensure(result === 17, expression)
}

function testExpression5() {
  let expression = "7 + 3 * (10 / (12 / (3 + 1) - 1))"
  let interpreter = makeInterpreter(expression)
  let result = interpreter.interpret()
  ensure(result === 22, expression)
}

function testExpression6() {
  let expression = "7 + 3 * (10 / (12 / (3 + 1) - 1)) / (2 + 3) - 5 - 3 + (8)"
  let interpreter = makeInterpreter(
    expression,
  )
  let result = interpreter.interpret()
  ensure(result === 10, expression)
}

function testExpression7() {
  let expression = "7 + (((3 + 2)))"
  let interpreter = makeInterpreter(expression)
  let result = interpreter.interpret()
  ensure(result === 12, expression)
}

function testExpression8() {
  let expression = "10 *"
  let interpreter = makeInterpreter(expression)
  let result = interpreter.interpret()
  ensure(result === 0, expression)
}
function main() {
  // let text = `(3 + 5) * 6`
  // let lexer = new Lexer(text)
  // let parser = new Parser(lexer)
  // let tree = parser.parse()
  // let interpreter = new Interpreter(tree)
  // let result = interpreter.interpret()
  // console.log('interpret ->', result)

  testExpression1()
  testExpression2()
  testExpression3()
  testExpression4()
  testExpression5()
  testExpression6()
  testExpression7()
  // testExpression8()

}

main()

File: lsbasi/part8/go/interpreter.go
package main

import (
	"fmt"
	"log"
	"regexp"
	"strconv"
	"unicode/utf8"
)

type Token struct {
	tokenType string
	value     string
}

func NewToken(tokenType string, value string) *Token {
	return &Token{
		tokenType: tokenType,
		value:     value,
	}
}

func (token *Token) format() string {
	s := "{ " + token.tokenType + " " + token.value + " }"
	return s
}

//
// LEXER
//

const (
	INTEGER = "INTEGER"
	PLUS    = "PLUS"
	MINUS   = "MINUS"
	MUL     = "MUL"
	DIV     = "DIV"
	MOD     = "MOD"
	POWER   = "POWER"
	LPAREN  = "LPAREN"
	RPAREN  = "RPAREN"
	EOF     = "EOF"
)

type Lexer struct {
	text        string
	pos         int
	currentChar *string
}

func NewLexer(text string) *Lexer {
	pos := 0
	currentChar := string([]rune(text)[pos])
	return &Lexer{
		text:        text,
		pos:         pos,
		currentChar: &currentChar,
	}
}

func (lexer *Lexer) format() string {
	s := "{ text: " + lexer.text +
		", pos: " + strconv.Itoa(lexer.pos) +
		", currentChar: " + *lexer.currentChar + " }"
	return s
}

func (lexer *Lexer) advance() {
	lexer.pos++
	if lexer.pos > utf8.RuneCountInString(lexer.text)-1 {
		lexer.currentChar = nil
	} else {
		*lexer.currentChar = string([]rune(lexer.text)[lexer.pos])
	}
}

func isSpace(word string) bool {
	whitespace := regexp.MustCompile(`\s`).MatchString(word)
	return whitespace
}

func (lexer *Lexer) skipWhitespace() {
	if lexer.currentChar != nil && isSpace(*lexer.currentChar) {
		lexer.advance()
	}
}

func isInteger(word string) bool {
	integer := regexp.MustCompile(`\d`).MatchString(word)
	return integer
}

func (lexer *Lexer) integer() string {
	result := ""

	for lexer.currentChar != nil && isInteger(*lexer.currentChar) {
		result += *lexer.currentChar
		lexer.advance()
	}
	// n, _ := strconv.Atoi(result)
	return result
}

func (lexer *Lexer) getNextToken() *Token {

	for lexer.currentChar != nil {

		if isSpace(*lexer.currentChar) {
			lexer.skipWhitespace()
			continue
		}

		if isInteger(*lexer.currentChar) {
			return NewToken(INTEGER, lexer.integer())
		}

		if *lexer.currentChar == "+" {
			lexer.advance()
			return NewToken(PLUS, "+")
		}

		if *lexer.currentChar == "-" {
			lexer.advance()
			return NewToken(MINUS, "-")
		}

		if *lexer.currentChar == "*" {
			lexer.advance()
			return NewToken(MUL, "*")
		}

		if *lexer.currentChar == "/" {
			lexer.advance()
			return NewToken(DIV, "/")
		}

		if *lexer.currentChar == "^" {
			lexer.advance()
			return NewToken(POWER, "^")
		}

		if *lexer.currentChar == "%" {
			lexer.advance()
			// fmt.Println("MOD ->", MOD)
			return NewToken(MOD, "%")
		}

		if *lexer.currentChar == "(" {
			lexer.advance()
			return NewToken(LPAREN, "(")
		}

		if *lexer.currentChar == ")" {
			lexer.advance()
			return NewToken(RPAREN, ")")
		}

	}
	return NewToken(EOF, "nil")
}

//
// Parser
//

type AST struct {
	token    *Token
	astType  string
	children []AST
}

// astType
const (
	UNARY_OP = "UNARY_OP"
	NUM      = "NUM"
	BIN_OP   = "BIN_OP"
)

func NewAST(token *Token, astType string, children []AST) *AST {
	return &AST{
		token:    token,
		astType:  astType,
		children: children,
	}
}

type Parser struct {
	lexer        *Lexer
	currentToken *Token
}

func NewParser(lexer *Lexer) *Parser {
	currentToken := lexer.getNextToken()
	return &Parser{
		lexer:        lexer,
		currentToken: currentToken,
	}
}

func (parser *Parser) error() {
	log.Fatal("Invalid syntax")
}

func (parser *Parser) eat(tokenType string) {
	if parser.currentToken.tokenType == tokenType {
		parser.currentToken = parser.lexer.getNextToken()
	} else {
		parser.error()
	}
}

func (parser *Parser) factor() *AST {
	token := parser.currentToken

	if parser.currentToken.tokenType == PLUS {
		parser.eat(PLUS)
		node := NewAST(token, UNARY_OP, []AST{*parser.factor()})
		return node
	}

	if parser.currentToken.tokenType == MINUS {
		parser.eat(MINUS)
		node := NewAST(token, UNARY_OP, []AST{*parser.factor()})
		return node
	}

	if token.tokenType == INTEGER {
		parser.eat(INTEGER)
		return NewAST(token, NUM, []AST{})
	}

	if token.tokenType == LPAREN {
		parser.eat(LPAREN)
		node := parser.expr()
		parser.eat(RPAREN)
		return node
	}
	return &AST{}
}

func (parser *Parser) term() *AST {
	node := parser.factor()

	for parser.currentToken.tokenType == MUL ||
		parser.currentToken.tokenType == DIV ||
		parser.currentToken.tokenType == MOD {
		token := parser.currentToken

		if parser.currentToken.tokenType == MUL {
			parser.eat(MUL)
			children := []AST{*node, *parser.factor()}
			node = NewAST(token, BIN_OP, children)

		} else if parser.currentToken.tokenType == DIV {
			parser.eat(DIV)
			children := []AST{*node, *parser.factor()}
			node = NewAST(token, BIN_OP, children)

		} else if parser.currentToken.tokenType == MOD {
			parser.eat(MOD)
			children := []AST{*node, *parser.factor()}
			node = NewAST(token, BIN_OP, children)
		}
	}
	return node
}

func (parser *Parser) expr() *AST {
	node := parser.term()

	for parser.currentToken.tokenType == PLUS ||
		parser.currentToken.tokenType == MINUS {
		token := parser.currentToken

		if parser.currentToken.tokenType == PLUS {
			parser.eat(PLUS)
			children := []AST{*node, *parser.term()}
			node = NewAST(token, BIN_OP, children)

		} else if parser.currentToken.tokenType == MINUS {

			parser.eat(MINUS)
			children := []AST{*node, *parser.term()}
			node = NewAST(token, BIN_OP, children)
		}
	}
	return node
}

func (parser *Parser) parse() *AST {
	node := parser.expr()
	if parser.currentToken.tokenType != EOF {
		fmt.Println("current token ->", parser.currentToken)
		parser.error()
	}
	return node
}

func (parser *Parser) each() {
	for parser.currentToken.tokenType != EOF {
		fmt.Println("current token ->", parser.currentToken)
		parser.currentToken = parser.lexer.getNextToken()
	}
}

type Interpreter struct {
	parser *Parser
}

func NewInterpreter(parser *Parser) *Interpreter {
	interpreter := Interpreter{
		parser: parser,
	}
	return &interpreter
}

func (interpreter *Interpreter) visitNum(node *AST) int {
	n, _ := strconv.Atoi(node.token.value)
	return n
}

func (interpreter *Interpreter) visitBinop(node *AST) int {

	leftVal := interpreter.visit(&node.children[0])
	rightVal := interpreter.visit(&node.children[1])

	if node.token.tokenType == PLUS {
		return leftVal + rightVal
	}
	if node.token.tokenType == MINUS {
		return leftVal - rightVal
	}
	if node.token.tokenType == MUL {
		return leftVal * rightVal
	}
	if node.token.tokenType == DIV {
		return leftVal / rightVal
	}
	if node.token.tokenType == MOD {
		return leftVal % rightVal
	}
	log.Fatal("Visit Binop Error")
	return 0
}

func (interpreter *Interpreter) visitUnaryOp(node *AST) int {
	if node.token.tokenType == PLUS {
		return +interpreter.visit(&node.children[0])
	}
	if node.token.tokenType == MINUS {
		return -interpreter.visit(&node.children[0])
	}

	log.Fatal("Visit UnaryOp Error")
	return 0
}

func (interpreter *Interpreter) visit(node *AST) int {

	if node.astType == UNARY_OP {
		return interpreter.visitUnaryOp(node)
	}

	if node.astType == BIN_OP {
		return interpreter.visitBinop(node)
	}

	if node.astType == NUM {
		return interpreter.visitNum(node)
	}
	log.Fatal("Visit Error")
	return 0
}

func (interpreter *Interpreter) interpret() int {
	tree := interpreter.parser.parse()
	result := interpreter.visit(tree)
	return result
}

func ensure(condition bool, message string) {
	if !condition {
		fmt.Println("*** 测试失败:", message)
	} else {

		fmt.Println("*** 测试成功:", message)
	}
}

func makeInterpreter(text string) *Interpreter {
	lexer := NewLexer(text)
	parser := NewParser(lexer)
	interpreter := NewInterpreter(parser)
	return interpreter
}

func testExpression1() {
	interpreter := makeInterpreter("3")
	result := interpreter.interpret()
	ensure(result == 3, "test_expression1")
}

func testExpression2() {
	interpreter := makeInterpreter("2 + 7 * 4")
	result := interpreter.interpret()
	ensure(result == 30, "test_expression2")
}

func testExpression3() {
	interpreter := makeInterpreter("7 - 8 / 4")
	result := interpreter.interpret()
	ensure(result == 5, "test_expression3")
}

func testExpression4() {
	interpreter := makeInterpreter("14 + 2 * 3 - 6 / 2")
	result := interpreter.interpret()
	ensure(result == 17, "test_expression4")
}

func testExpression5() {
	interpreter := makeInterpreter("7 + 3 * (10 / (12 / (3 + 1) - 1))")
	result := interpreter.interpret()
	ensure(result == 22, "test_expression5")
}

func testExpression6() {
	interpreter := makeInterpreter(
		"7 + 3 * (10 / (12 / (3 + 1) - 1)) / (2 + 3) - 5 - 3 + (8)",
	)
	result := interpreter.interpret()
	ensure(result == 10, "test_expression6")
}

func testExpression7() {
	interpreter := makeInterpreter("7 + (((3 + 2)))")
	result := interpreter.interpret()
	ensure(result == 12, "test_expression7")
}

func testExpression8() {
	interpreter := makeInterpreter("10 *")
	result := interpreter.interpret()
	ensure(result == 0, "test_expression8")
}

func main() {
	// lexer := NewLexer("8 / -8 % 6")
	// parser := NewParser(lexer)
	// parser.each()
	// interpreter := NewInterpreter(parser)
	// r := interpreter.interpret()
	// fmt.Println("r ->", r)
	// fmt.Println("r ->", 8/-8%6)
	testExpression1()
	testExpression2()
	testExpression3()
	testExpression4()
	testExpression5()
	testExpression6()
	testExpression7()
	testExpression8()
}
